### PR TITLE
Leader scheduler plumbing

### PR DESCRIPTION
**Pull request by carllin**
Integrated LeaderScheduler with:

Fullnode - Check leader schedule on startup after processing ledger, and pick a role based on that schedule.

Bank - Process ledger/blocks/entry functions now update the LeaderScheduler, so the leader schedule is known on startup and leader <-> validator transitions

Tvu - Integrated LeaderScheduler with ReplicateStage, removed leader rotation checks from WindowService to get around problem where the WindowService wouldn't know how to calculate stakes for leader rotation at a particular entry height, b/c ReplicateStage hasn't run the necessary transactions through the bank for that entry height.

Tpu - Integrated LeaderScheduler with WriteStage, remove unnecessary leader checks from Broadcast stage because WriteStage will stop sending entries at the correct point.

Removed old update_leader / leader scheduler functions from Crdt.

Added integration tests in multinode.

### CODE REVIEW

File: src/leader_scheduler.rs [hunk 1]
```diff
@@ -769,18 +844,15 @@ mod tests {
         for i in 0..=num_validators {
             leader_scheduler.generate_schedule(i * active_window_length, &bank);
             let result = &leader_scheduler.leader_schedule;
-            let mut expected_set;
+            let mut expected;
```
Comment (garious) [inline src/leader_scheduler.rs]: Please get rid of the muts when you see the opportunity. How about?

```rust
let expected = if i == num_validators { bootstram_leader_id } else {validators[i as usize]};
assert_eq!(vec![expected], *result);
```

File: src/fullnode.rs [hunk 2]
```diff
@@ -702,21 +865,17 @@ mod tests {
...
-        let (_, entry_height, _) = Fullnode::new_bank_from_ledger(&validator_ledger_path);
+        // Check the validator ledger to make sure it's the right height, we should've
+        // transitioned after the bootstrap_height entry
+        let (_, entry_height, _) = Fullnode::new_bank_from_ledger(&validator_ledger_path, None);
 
-        assert_eq!(
-            entry_height,
-            my_leader_begin_epoch * leader_rotation_interval
-        );
+        assert_eq!(entry_height, bootstrap_height,);
```
Comment (garious) [inline src/fullnode.rs]: Extra comma

File: src/hash.rs [hunk 3]
```diff
@@ -7,7 +7,7 @@ use sha2::{Digest, Sha256};
 use std::fmt;
 
 #[derive(Serialize, Deserialize, Clone, Copy, Default, Eq, PartialEq, Ord, PartialOrd, Hash)]
-pub struct Hash(GenericArray<u8, U32>);
+pub struct Hash(pub GenericArray<u8, U32>);
```
Comment (garious) [inline src/hash.rs]: Again? How'd this happen?

File: src/leader_scheduler.rs [hunk 4]
```diff
@@ -322,6 +335,77 @@ impl LeaderScheduler {
...
+pub fn is_leader_rotation_height(
+    height: u64,
+    bootstrap_height: u64,
+    leader_rotation_interval: u64,
+) -> bool {
+    if height < bootstrap_height {
+        return false;
+    }
+
+    if (height - bootstrap_height) % leader_rotation_interval == 0 {
```
Comment (garious) [inline src/leader_scheduler.rs]: Redundant if-else-return-false. Reduce to: `(height - bootstrap_height) % leader_rotation_interval == 0`

File: src/replicate_stage.rs [hunk 5]
```diff
@@ -129,21 +204,127 @@ impl ReplicateStage {
...
+        // Create a ledger
+        let (mint, my_ledger_path) = genesis("test_replicate_stage_leader_rotation_exit", 10_000);
+        let genesis_entries = mint.create_entries();
+        let mut last_id = genesis_entries
+            .last()
+            .expect("expected at least one genesis entry")
+            .id;
+
+        // Write two entries to the ledger so that the validator is in the active set:
+        // 1) Give him nonzero number of tokens 2) A vote from the validator .
```
Comment (garious) [inline src/replicate_stage.rs]: Please avoid pronouns in the codebase, especially ones with gender.

File: src/window.rs [hunk 6]
```diff
@@ -101,13 +104,54 @@ impl WindowUtil for Window {
...
+                "leader_scheduler_option cannot be None if 
+                leader_rotation_interval_option is not None",
+            );
+            match leader_scheduler
+                .read()
+                .unwrap()
+                .get_scheduled_leader(next_leader_rotation)
+            {
+                Some(leader_id) if leader_id == *id => is_next_leader = true,
+                // In the case that we are not in the current scope of the leader schedule
```
Comment (garious) [inline src/window.rs]: Seems like this kind of thing would be especially easy for a newcomer to break by accident. Can these comments be replaced by commented unit-tests?

File: src/banking_stage.rs [hunk 7]
```diff
@@ -85,7 +85,6 @@ impl BankingStage {
 
         // Many banks that process transactions in parallel.
         let mut thread_hdls: Vec<JoinHandle<()>> = (0..NUM_THREADS)
-            .into_iter()
```
Comment (rob-solana) [inline src/banking_stage.rs]: this is redundant, a clippy warning?

File: src/broadcast_stage.rs [hunk 8]
```diff
@@ -183,38 +187,29 @@ impl BroadcastStage {
         window: &SharedWindow,
         entry_height: u64,
         receiver: &Receiver<Vec<Entry>>,
+        leader_scheduler_option: &Option<Arc<RwLock<LeaderScheduler>>>,
```
Comment (rob-solana) [inline src/broadcast_stage.rs]: is the _option suffix a convention we're starting, already using?  where does this come from?

File: src/leader_scheduler.rs [hunk 9]
```diff
@@ -322,6 +335,77 @@ impl LeaderScheduler {
...
+    let mut accounts = bank.accounts().write().unwrap();
+    accounts.clear();
+    accounts.insert(new_leader_id, new_account);
+}
+
+// Create two entries so that the node with keypair == active_keypair
+// is in the active set for leader selection:
+// 1) Give him nonzero number of tokens,
+// 2) A vote from the validator
+pub fn make_active_set_entries(
```
Comment (rob-solana) [inline src/leader_scheduler.rs]: this also a test function?

File: src/replicate_stage.rs [hunk 10]
```diff
@@ -59,27 +68,77 @@ impl ReplicateStage {
             entries.append(&mut more);
         }
 
-        let res = bank.process_entries(&entries);
+        let mut res = Ok(());
+        {
```
Comment (rob-solana) [inline src/replicate_stage.rs]: what if process_entry() returned the stuff we need at this level, instead?

Comment (rob-solana) [reply src/replicate_stage.rs]: all the options makes for quite a mess, ey?

File: src/broadcast_stage.rs (hunk 8 above)
Comment (garious) [reply src/broadcast_stage.rs]: It's new. Traditionally we've omitted any distinction between `Option<T>` and `T`. If we're going to distinguish these, I'd prefer someone poke around for a community convention first.

Comment (rob-solana) [reply src/broadcast_stage.rs]: my personal preference would be to not make a distinction

File: src/leader_scheduler.rs (hunk 9 above)
Comment (carllin) [reply src/leader_scheduler.rs]: yup! but it's also used by the integration tests in multinode.rs, so I couldn't mark it with the "test" annotation

File: src/banking_stage.rs (hunk 7 above)
Comment (carllin) [reply src/banking_stage.rs]: yup!

File: src/bank.rs [hunk 11]
```diff
@@ -802,13 +821,30 @@ impl Bank {
                 return Err(BankError::LedgerVerificationFailed);
             }
             id = block.last().unwrap().id;
-            entry_count += self.process_entries_tail(block, tail, tail_idx)?;
+            let tail_count = self.process_entries_tail(&block, tail, tail_idx)?;
+
+            for (i, entry) in block.iter().enumerate() {
+                if !leader_scheduler.use_only_bootstrap_leader {
```
Comment (rob-solana) [inline src/bank.rs]: use_only_bootstrap_leader looks invariant for this loop, can you lift the check?

**Comment by garious** [comment]
@rob-solana, @carllin, merge it?

**Pull request merged**